doc/ACL/SAI-Proposal-Exact-Match.md

```diff
@@ -0,0 +1,123 @@
+- There is no concept of entry ordering or priority with exact match. All entries need to have the same priority. The implementation can reject entries if there is a mix of priority values.
+- Masking selected bits in an ACL entry is not supported with exact match since the table is generally implemented using hash lookup. Implementation can reject entry add requests if the mask is not all 1s.
+- If all bits in a field are not required to be part of the lookup key, the valid bits can be specified at table level. Every ACL entry in the table will use the valid bit specification to determine the lookup key.
+
+# ACL Specification Enhancements #
+The proposal includes following changes to the ACL spec.
+
+1. Define a new ACL stage named SAI_ACL_STAGE_POST_INGRESS which is placed after router table lookup in the pipeline. The new ACL stage allows setting of user metadata from SAI_ACL_STAGE_INGRESS and matching the user metadata in SAI_ACL_STAGE_POST_INGRESS.
+
+The post ingress ACL stage packet action will overwrite the action from previous stages if there are conflicting packet actions. For example, a redirect action from post ingress stage will take priority over a redirect action from ACL ingress stage and the next hop id from router table.
+
+2. Define a new table attribute  SAI_ACL_TABLE_ATTR_ACL_MATCH_TYPE of type sai_acl_match_t to identify an ACL table match type.
+```
+typedef enum _sai_acl_match_type_t
+{
+    /** ternary matching */
+    SAI_ACL_MATCH_TYPE_TERNARY,
+
+    /** exact matching */
+    SAI_ACL_MATCH_TYPE_EXACT,
+
+} sai_acl_match_t;
+```
+
+3. Add capability enhancments to query SDK to determine the match types that an ACL stage can support.
+```
+/**
+ * @brief Attribute data for SAI_ACL_TABLE_ATTR_SUPPORTED_MATCH_TYPE
+ */
+typedef enum _sai_acl_supported_match_type_t
+{
+    /** Ternary */
+    SAI_ACL_SUPPORTED_MATCH_TYPE_TERNARY,
+
+    /** Exact */
+    SAI_ACL_SUPPORTED_MATCH_TYPE_EXACT,
+
+    /** Ternary and Exact */
+    SAI_ACL_SUPPORTED_MATCH_TYPE_TERNARY_AND_EXACT,
+
+} sai_acl_supported_match_type_t;
+
+typedef struct _sai_acl_capability_t
+{
+    :
+    /**
+     * @brief Supported match type
+     *
+     * Indicates the match type supported by ACL stage
+     */
+    sai_acl_supported_match_type_t supported_match_type;
+
+} sai_acl_capability_t;
+```
+
+4. Add new attributes to ACL table to specify valid bits for keys.  If a match field does not have valid bits specified, all bits in the field are valid. For tables implemented using Exact Match, there is no further key masking supported in table entries. The mask for such entries needs to be set to all 1s.
+```
+    /**
+     * @brief Src IPv6 Valid bits
+     *
+     * @type sai_acl_field_data_mask_t sai_ip6_t
+     * @flags CREATE_ONLY
+     * @default ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff
+     * @validonly SAI_ACL_TABLE_ATTR_FIELD_SRC_IPV6 == true
+     */
+    SAI_ACL_TABLE_ATTR_FIELD_VALID_BITS_SRC_IPV6 = SAI_ACL_TABLE_ATTR_FIELD_VALID_BITS_START,
```

Comment:
From the description above, it says "mask for such entries need to be set to all 1s".  If there's no other choice to set this, what's the need of this field?  If user pass it a value with mixed of 1 or 0, is it allowed, or how should SAI adaptor handle it?

Reply:
The comment is referring to mask in acl entry. The mask field is needed in acl entry because it is a common structure shared by EM and TCAM ACLs. Ability to do mix of values is a capability which is added in PR1730 - https://github.com/opencomputeproject/SAI/pull/1730/files#diff-444febffce442e4277a1ced22d1ab94df045dfaa10a1b671ee35a7e5c2d9f683R836